messybrainz/create_recording_clusters.py

```diff
@@ -0,0 +1,208 @@
+        return False
+
+
+def insert_recording_cluster(connection, cluster_id, recording_gids):
+    """Creates new cluster in the recording_cluster table.
+
+    Args:
+        connection: the sqlalchemy db connection to be used to execute queries
+        cluster_id (UUID): the recording MSID which will represent the cluster.
+        recording_gids (list): the list of MSIDs will form a cluster.
+
+    Returns:
+        True if a new value is added to cluster otherwise False is returned.
+    """
+
+    query = text("""INSERT INTO recording_cluster (cluster_id, recording_gid, updated)
+                         VALUES (:cluster_id, :recording_gid, now())
+    """)
+    cluster_modified = False
+    for recording_gid in recording_gids:
+        if not is_recording_cluster_present_in_recording_cluster(connection, cluster_id, recording_gid):
+            connection.execute(query, {
+                "cluster_id": cluster_id,
+                "recording_gid": recording_gid,
+                }
+            )
+            cluster_modified = True
+    return cluster_modified
+
+
+def fetch_gids_for_recording_mbid(connection, recording_mbid):
+    """Fetches the gids corresponding to a recording_mbid.
+
+    Args:
+        connection: the sqlalchemy db connection to be used to execute queries
+        recording_mbid (UUID): the recording MBID for which gids are to be fetched.
+
+    Returns:
+        List of gids.
+    """
+
+    query = text("""SELECT gid
+                      FROM recording_json
+                      JOIN recording
+                        ON recording_json.id = recording.data
+                     WHERE recording_json.data ->> 'recording_mbid' = :recording_mbid
+    """)
+    gids = connection.execute(query, {
+        "recording_mbid": recording_mbid,
+        }
+    )
+    return [gid[0] for gid in gids]
+
+
+def fetch_distinct_recording_mbids(connection):
+    """Fetch all the distinct recording MBIDs we have in recording_json table.
+
+    Args:
+        connection: the sqlalchemy db connection to be used to execute queries
+
+    Returns:
+        recording_mbids.
+    """
+
+    query = text("""SELECT DISTINCT data ->> 'recording_mbid' AS recording_mbids
+                               FROM recording_json 
+                              WHERE data ->> 'recording_mbid' IS NOT NULL
+                """)
+    recording_mbids = connection.execute(query)
+    return recording_mbids
+
+
+def link_recording_mbid_to_recording_msid(connection, cluster_id, mbid):
+    """Links the recording mbid to the cluster_id.
+
+    Args:
+        connection: the sqlalchemy db connection to be used to execute queries
+        cluster_id: the gid which represents the cluster.
+        mbid: mbid for the cluster.
+
+    Returns:
+        True if link is inserted into the cluster otherwise False is returned.
+    """
+
+    cluster_added = False
+    if not is_recording_cluster_present_in_recording_redirect(connection, cluster_id, mbid):
```

Comment:
There is a better way to do this that to query the data and then take action if the data isn't there. Namely adding a unique table constraint that allows only one unique recording_cluster_id, recording_id pair to be added to the table. Then you simply carry out the insertion -- if the data already exists it will throw an error that you can safely ignore. If the data isn't there, it will be inserted correctly.